### PR TITLE
Consolidating parameters in GAC costing model

### LINKED ISSUE #975: Consolidate amount of costing parameter variables in GAC costing
### Description

With the new configuration option in #960, costing parameters of the unused option are constructed but unused. 

### Motivation

Effort to provide concise model code and reporting in `.display()`.

### Possible Implementation

- Index `Var()` based on config options since equations are of the same form
- If conditional to construct `Var()` based on which config option is selected

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by hunterbarber**
## Fixes/Resolves:
In response to issue #975

## Summary/Motivation:
The original costing model constructed 20 costing parameters, 11 of which were used in calculations dependent on the configuration argument. All parameters had their own Pyomo variables are therefore cluttered the code and reporting in `.display()`.

## Changes proposed in this PR:
- Create utility `_unpack_data_dict` to transform nested dictionaries into readable initialization objects by `pyo.Var()` (currently located within `watertap/costing/units/gac.py`)
- Convert individual Pyomo variables to be more concise
- With the given implementation, the need for rules based on configuration options are unnecessary since the variables are indexed by the configuration argument, therefore remove these

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by hunterbarber** [comment]
@adam-a-a Take a look and let me know what you think for this formatting of parameters before I update the documentation and request reviews

File: watertap/costing/units/gac.py
Comment (bknueven) [inline watertap/costing/units/gac.py]: Why make this change? I think it was better before ...

File: watertap/unit_models/tests/test_gac.py
Comment (bknueven) [inline watertap/unit_models/tests/test_gac.py]: The whole point of the `StrEnum` is you don't have to import these enums everywhere to use them. As an example I like the prior code better.

**Review by bknueven — COMMENTED** [review]
I have a few minor comments. I would like to merge this soon-ish so I can finally finish #676.

File: watertap/unit_models/tests/test_gac.py
Comment (hunterbarber) [reply watertap/unit_models/tests/test_gac.py]: This is fine to change back, I made it just to have my tests passing as I was playing around with changing names of the `StrEnum` options.

File: watertap/costing/units/gac.py
Comment (hunterbarber) [reply watertap/costing/units/gac.py]: I made this change just to have equal string lengths in the main code as I was building dictionaries and it would make it visually consistent. Bit of an extreme formatting thing. This also would be fine to change back.

Comment (adam-a-a) [reply watertap/costing/units/gac.py]: I'd prefer to change it back too for the sake of clarity.

**Comment by hunterbarber** [comment]
> I have a few minor comments. I would like to merge this soon-ish so I can finally finish #676.

@bknueven If you think this is a formatting improvement this can come out of draft, I'll revert those naming changes as well. The question I have is is this the best option of these two to consolidate unused costing parameters in this case?

1. (Current change in this PR) All the Pyomo variables that are constructed are used, however these variables are now indexed where only the index corresponding to the specified configuration option is used. Technically, the other index of the variable still contains unused parameters, they just don't overflow the reporting.
2. Add a conditional to only construct the variables when the configuration option is selected. This would increase the code length slightly (from the current change in this PR) but it would remove the unused parameters entirely.

**Comment by bknueven** [comment]
> > I have a few minor comments. I would like to merge this soon-ish so I can finally finish #676.
> 
> @bknueven If you think this is a formatting improvement this can come out of draft, I'll revert those naming changes as well. The question I have is is this the best option of these two to consolidate unused costing parameters in this case?
> 
>     1. (Current change in this PR) All the Pyomo variables that are constructed are used, however these variables are now indexed where only the index corresponding to the specified configuration option is used. Technically, the other index of the variable still contains unused parameters, they just don't overflow the reporting.
> 
>     2. Add a conditional to only construct the variables when the configuration option is selected. This would increase the code length slightly (from the current change in this PR) but it would remove the unused parameters entirely.

Thanks for the explanation. I prefer option (2) for clarity when printing the model and examining it in code. IDAES/WaterTAP conditionally build variables / constraints all the time and option (2) conforms better to that convention.

**Comment by hunterbarber** [comment]
> > > I have a few minor comments. I would like to merge this soon-ish so I can finally finish #676.
> > 
> > 
> > @bknueven If you think this is a formatting improvement this can come out of draft, I'll revert those naming changes as well. The question I have is is this the best option of these two to consolidate unused costing parameters in this case?
> > ```
> > 1. (Current change in this PR) All the Pyomo variables that are constructed are used, however these variables are now indexed where only the index corresponding to the specified configuration option is used. Technically, the other index of the variable still contains unused parameters, they just don't overflow the reporting.
> > 
> > 2. Add a conditional to only construct the variables when the configuration option is selected. This would increase the code length slightly (from the current change in this PR) but it would remove the unused parameters entirely.
> > ```
> 
> Thanks for the explanation. I prefer option (2) for clarity when printing the model and examining it in code. IDAES/WaterTAP conditionally build variables / constraints all the time and option (2) conforms better to that convention.

Sure, I'll implement that and combine with the existing improvements. I'm out the next two work days, but it can be done next week.

File: watertap/costing/units/gac.py
Comment (adam-a-a) [inline watertap/costing/units/gac.py]: The USD_2020 comment can go above to avoid any funky black formatting that may arise

File: watertap/costing/units/gac.py
Comment (adam-a-a) [inline watertap/costing/units/gac.py]: Maybe this should be in terms of the length of the associated dictionary.

**Comment by hunterbarber** [comment]
> > > > I have a few minor comments. I would like to merge this soon-ish so I can finally finish #676.
> > > 
> > > 
> > > @bknueven If you think this is a formatting improvement this can come out of draft, I'll revert those naming changes as well. The question I have is is this the best option of these two to consolidate unused costing parameters in this case?
> > > ```
> > > 1. (Current change in this PR) All the Pyomo variables that are constructed are used, however these variables are now indexed where only the index corresponding to the specified configuration option is used. Technically, the other index of the variable still contains unused parameters, they just don't overflow the reporting.
> > > 
> > > 2. Add a conditional to only construct the variables when the configuration option is selected. This would increase the code length slightly (from the current change in this PR) but it would remove the unused parameters entirely.
> > > ```
> > 
> > 
> > Thanks for the explanation. I prefer option (2) for clarity when printing the model and examining it in code. IDAES/WaterTAP conditionally build variables / constraints all the time and option (2) conforms better to that convention.
> 
> Sure, I'll implement that and combine with the existing improvements. I'm out the next two work days, but it can be done next week.

@bknueven @adam-a-a  The current file conforms to not constructing any unused variables for all indices. Had to do a little niche method in `cost_gac` to have it working as intended, but the in-line comments should cover the reasoning. I double-checked the variables built on the model and the tests, so everything should be good as long as this niche method isn't flawed in some way. Just let me know if you find something. (With the okay on this method I also have to change the docs before its merged.)

File: watertap/costing/units/gac.py
Comment (bknueven) [inline watertap/costing/units/gac.py]: It's a small thing, but you could use `ComponentMap` here to map the var to the coeff data.

Alternatively, you could create a similar `tuple`, e.g., `[( blk.costing_package.gac.adsorbent_unit_cost_coeff, adsorbent_unit_cost_coeff_data), ...]`.

Or, we could use `zip` to iterate over the list simultaneously. This is my least favorite option. Ideally the var would be close to the coeff_data in the code.

File: watertap/costing/units/gac.py
Comment (bknueven) [inline watertap/costing/units/gac.py]: I would consider moving these temporary dictionaries close to where they're used.

File: watertap/costing/units/gac.py
Comment (bknueven) [inline watertap/costing/units/gac.py]: You can just pass in the dictionary directly, or use `.keys()` if you want to be explicit, to set the indices. E.g.,
```suggestion
        contactor_cost_coeff_dummy,
```

**Review by bknueven — APPROVED** [review]
Looks good; I have a few Python stylistic suggestions

File: watertap/costing/units/gac.py
Comment (hunterbarber) [reply watertap/costing/units/gac.py]: I thought the same thing originally with the tuple, but figured it wouldn't actually consolidate anything because black would end up formatting it in a similar way. I will look into the `ComponentMap` briefly though.

Comment (bknueven) [inline watertap/costing/units/gac.py]: Following up from above. Maybe this is better than the `ComponentMap` idea because it's temporary.

```suggestion
    # iterable matching coeff_data to vars
    cost_params = (
        (adsorbent_unit_cost_coeff_data, blk.costing_package.gac.adsorbent_unit_cost_coeff),
        (contactor_cost_coeff_data, blk.costing_package.gac.contactor_cost_coeff),
        (other_cost_param_data, blk.costing_package.gac.other_cost_param),
        (energy_consumption_coeff_data, blk.costing_package.gac.energy_consumption_coeff),
    )

    # refix variables to appropriate costing parameters
    for data, var in cost_params:
        for index, var in param_var.items():
            var.fix(data[index])
```

Comment (bknueven) [reply watertap/costing/units/gac.py]: The point with the `tuples` is that it's easier to catch mistakes because the two pieces of information are together in the code. Same with `ComponentMap`.

**Comment by hunterbarber** [comment]
> Looks good; I have a few Python stylistic suggestions

@bknueven Made sense, and I applied those suggestions in the latest commit, thanks. As soon as these docs to reflect these changes are passing (there's been GitHub issues today so it may wait on that) this should be merged.

File: docs/technical_reference/unit_models/gac.rst
Comment (adam-a-a) [inline docs/technical_reference/unit_models/gac.rst:249]: Can you add the linked reference to the EPA source around the end of the doc?

**Review by adam-a-a — APPROVED** [review]
LGTM

File: docs/technical_reference/unit_models/gac.rst
Comment (hunterbarber) [reply docs/technical_reference/unit_models/gac.rst:249]: Sure, just added the link to the end of the ref.

**Comment by hunterbarber** [comment]
@bknueven @adam-a-a Thanks. This is ready to merge.

**Pull request merged**
